Plugin generator

alepuccetti opened the pull request:
This PR introduces  Weave Scope plugin generator's code and documentation.

/cc @2opremio

Commit:
core: initial code for plugin generator

This commit adds the basic files for the plugin generator.

Commit:
doc: add scope plugin generator documentation

Commit:
core: add pre-post generating hooks

`pre_gen_project.py` checks the validity of `plugin_id` and `docker_repo`,
`post_gen_project.sh` sets up the git local git repository and add the build-tools
from https://github.com/weaveworks/build-tools

Fixes https://github.com/weaveworks-plugins/scope-plugin-generator/issues/1

Commit:
doc: add section about plugin generation result

This patch explains what is the result of the plugin generator.

alepuccetti commented:
This is relevant to https://github.com/weaveworks/scope/issues/1963

2opremio commented on README.md:
```diff
@@ -1,2 +1,64 @@
-# scope-plugin-generator
-This is a plugin generator for [Weave Scope](https://github.com/weaveworks/scope)
+# Scope Plugin Generator
+The Scope Plugin Generator uses [Cookiecutter](https://github.com/audreyr/cookiecutter) to automatically generate a [Weave Scope](https://github.com/weaveworks/scope) plugin's basic structure.
```
In what language? What can the user expect for the skeleton?

alepuccetti replied:
What to expect is in the following section. Do you want me to move it up here?

2opremio commented on README.md:
```diff
@@ -1,2 +1,64 @@
+To generate your own plugin you will need Cookiecutter, you can install it following these [instructions](https://cookiecutter.readthedocs.io/en/latest/installation.html).
+
+After you have installed Cookiecutter, you can create a basic Scope plugin structure using the following command: `cookiecutter https://github.com/weaveworks-plugins/scope-plugin-generator`.
+You will be prompted to insert the following basic information about your plugin:
```
Are all this all necessary? They seem a bit overwhelming.

If the intention is to get a fast start for users I would only require the absolutely fundamental options and then add an extra section indicating what they may possible when defined.

Also, please see if you can reduce their number.

alepuccetti replied:
I can do multiple incremental templates one that generate the code, one for the dockerfile and makefile.

2opremio commented on {{cookiecutter.project_root_dir_name}}/circle.yml:
```diff
@@ -0,0 +1,53 @@
+# You need to setup the DOCKER_USER, DOCKER_EMAIL, and DOCKER_PASS
```
Forcing users to use circle seems like a hard requirement.

alepuccetti replied:
They can use whatever, it seemed nice to provide a boilerplate for the CI/CD too. It is not a requirement, it is more an added bonus.

Commit:
core,doc: remove not used attributes

This patch removes some unused attributes. Also, it drops a couple of attributes
to semplify the steps for the user.

Commit:
doc: split mandatory and optional attributes

This patch clarifies which attributes are mandatory and which are optional.
However, the optional attributes are still filled with the default values.

Commit:
core: print the report in JSON format sent by the plugin

Although, the prints are commented so the plugin will not print the report
by default.

alepuccetti commented:
@2opremio I address your comments. I removed some attributes that are not essential, and I clarified in the doc the mandatory attributes and the optional ones.
I dropped the multiple templates approach because it was more confusing for the user. Now it is clear that they can just ignore the optional attributes and the `circle.yml` file.
